render/render_test.go

```diff
@@ -18,73 +18,159 @@ const (
 name: my-app`
 )
 
-func TestRenderMissingValue(t *testing.T) {
-	configFiles := []composetypes.ConfigFile{
-		{
-			Config: map[string]interface{}{
-				"version": "3",
-				"services": map[string]interface{}{
-					"foo": map[string]interface{}{
-						"image": "${imageName}:${version}",
-					},
-				},
-			},
-		},
+func TestSubstituteBracedParams(t *testing.T) {
+	composeFile := `
+version: "3.6"
+services:
+  front:
+    ports:
+     - "${front.port}:80"
+`
+	parameters := map[string]string{
+		"front.port": "8080",
 	}
-	finalEnv := map[string]string{
-		"imageName": "foo",
+	s, err := substituteParams(parameters, composeFile)
+	assert.NilError(t, err)
+	assert.Equal(t, s, `
+version: "3.6"
+services:
+  front:
+    ports:
+     - "8080:80"
+`)
+}
+
+func TestSubstituteNamedParams(t *testing.T) {
+	composeFile := `
+version: "3.6"
+services:
+  back:
+    ports:
+     - "$back.port:90"
+`
+	parameters := map[string]string{
+		"back.port": "9000",
 	}
-	_, err := render("foo.dockerapp", configFiles, finalEnv, nil)
-	assert.Check(t, err != nil)
-	assert.Check(t, is.ErrorContains(err, "required variable"))
+	s, err := substituteParams(parameters, composeFile)
+	assert.NilError(t, err)
+	assert.Equal(t, s, `
+version: "3.6"
+services:
+  back:
+    ports:
+     - "9000:90"
+`)
 }
 
-func TestRender(t *testing.T) {
-	configFiles := []composetypes.ConfigFile{
-		{
-			Config: map[string]interface{}{
-				"version": "3",
-				"services": map[string]interface{}{
-					"foo": map[string]interface{}{
-						"image":   "busybox:${version}",
-						"command": []interface{}{"-text", "${foo.bar}"},
-					},
-				},
-			},
-		},
+func TestSubstituteMixedParams(t *testing.T) {
+	composeFile := `
+version: "3.6"
+services:
+  front:
+    image: wordpress
+    ports:
+     - "${front.port}:80"
+    deploy:
+      replicas: ${front.deploy.replicas}
+  back:
+    image: mysql
+    ports:
+     - "$back.port:90"
+`
+	parameters := map[string]string{
+		"front.port":            "8080",
+		"back.port":             "9000",
+		"front.deploy.replicas": "3",
 	}
-	finalEnv := map[string]string{
-		"version": "latest",
-		"foo.bar": "baz",
+	s, err := substituteParams(parameters, composeFile)
+	assert.NilError(t, err)
+	assert.Equal(t, s, `
+version: "3.6"
+services:
+  front:
+    image: wordpress
+    ports:
+     - "8080:80"
+    deploy:
+      replicas: 3
+  back:
+    image: mysql
+    ports:
+     - "9000:90"
+`)
+}
+
+func TestSubstituteSkipDoubleDollarCase(t *testing.T) {
```

Comment:
It would be great if we simplifies this test case, the name says we should skip the double dollar sign but the compose file is quite big and it's not really obvious what is going on.